opencog/atoms/base/ClassServer.cc

```diff
@@ -189,21 +61,26 @@ static Handle validating_factory(const Handle& atom_to_check)
 
 void ClassServer::spliceFactory(Type t, AtomFactory* fact)
 {
+	// N.B. it is too late to synchronize calls with NameServer using a shared mutex.
+	// If registrations of class names interleave with registration of factories,
+	// then the code will not work properly anyway. Before a factory is registered,
+	// the complete class hierarchy must be known.
+	
```

Comment:
If this is really true, and can't be fixed, then this is a fatal design flaw. There are modules outside of the atomspace, and outside of opencog, that declare new atom types, and classes for them.

Reply:
The problem could occur if registration of class factories using ```addFactory``` precedes registration of _all_ type names. Because in such case factory methods might not be set properly for nested types, which are created after.

That might happen as the relative order of constructors execution is not guaranteed. A remedy is to define the priority in such a way, that names are always registered first. For instance, the following syntax might be used ```__attribute__((constructor (300)))```. Constructors with smaller numbers are executed first. This is weird and I am not sure, that there is an analogue for Windows.

A more general approach is to update ```_atomFactory``` each time a new type name is added. But this will look ugly too.

Reply:
I can guarantee that a type-name will be added before the factory for that type.  I can also guaranteee that some factories will be added before other type names are added.  For example, the core type-names will be added, and then their factories; then the NLP type names, and then their factories.  Its impossible to load the NLP type names before loading the core type factories.

Reply:
Well, even this guarantee might not give absolute reliability.

Here is an example. Imagine we have two atom types: Parent and Child. The parent has explicit factory, which should be implicitly inherited by the Child. But that will not happen in the following case:

```
Load library Parent:
- Register type-name Parent
- Register factory Parent. The following code executes

        // OOPS!, the child type does not exist⤵
	for (Type chi=t; chi < _nameServer.getNumberOfClasses(); chi++) 
	{
		if (_nameServer.isAncestor(t, chi) and
		    (nullptr == _atomFactory[chi] or
		    ok_to_clobber.end() != ok_to_clobber.find(_atomFactory[chi])))
		{
			_atomFactory[chi] = fact;
		}
	}

Load library Child:
- Register type-name Child
- The factory for child is left initialized

Reply:
As for NameServer-ClassSever concurrent execution, I will think on it, and likely will create one more pull request.

Reply:
I investigated this subject. Seems like there are a lot of implicit requirements on how to register types/factories:

1. No new types can be added to an existing module once its initialization is concluded using endTypeDecls. Otherwise declType will throw.
2. Initialization of modules must be completely serialized _by_ _clients_. I.e. it is not allowed to start initializing a new module before endTypeDecls is called for the previous module. Otherwise declType will throw. So no multi-threading can be used.
3. Class factory for an atom type can be registered only after all its children types are registered. Otherwise the children will not inherit the parent's factory.
4. If a type inherits two other, and both parents have non-trivial factories, then the child must have a factory explicitly declared. Otherwise the child will inherit one of parents' factories; which one - undefined.

These requirements are met everywhere in the repo, because constructors are called in just one thread, and execution order is specified by dependencies, defined in CMakeLists.txt (though, that is unspecified behavior of gcc). 
In such case NO serialization using mutexes is needed at all - the read access is safe.

>  its not uncommon that there might be 2 or 3 modules, all trying to load new atom types, and load factories for them, and this might be happening in different threads

I did not find such scenario. And I cannot see how that could work reliably with the present implementation (see above). @linas, please give a reference to a live multi-threading example.

P.S. maybe I dug an issue which cares nobody. If so, I could switch to another task.